openvdb/points/AttributeArrayString.h

```diff
@@ -49,27 +49,68 @@ struct StringCodec
 };
 
 
-using StringAttributeArray = TypedAttributeArray<StringIndexType, StringCodec<false>>;
+using StringAttributeArray = TypedAttributeArray<Index, StringCodec<false>>;
 
 
 ////////////////////////////////////////
 
 
+/// Class to compute a string->index map from all string:N metadata
+class OPENVDB_API StringMetaCache
+{
+public:
+    using UniquePtr = std::unique_ptr<StringMetaCache>;
+    using ValueMap = std::unordered_map<Name, Index>;
+
+    StringMetaCache() = default;
+    explicit StringMetaCache(const MetaMap& metadata);
+
+    /// Return @c true if no string elements in metadata
+    bool empty() const { return mCache.empty(); }
+    /// Returns the number of string elements in metadata
+    size_t size() const { return mCache.size(); }
+
+    /// Clears and re-populates the cache
+    void reset(const MetaMap& metadata);
+
+    /// Insert a new element in the cache
+    void insert(const Name& key, Index index);
+
+    /// Retrieve the value map (string -> index)
+    const ValueMap& map() const { return mCache; }
+
+private:
+    ValueMap mCache;
+}; // StringMetaCache
+
+
+////////////////////////////////////////
+
+
+/// Class to help with insertion of keyed string values into metadata
 class OPENVDB_API StringMetaInserter
 {
 public:
-    StringMetaInserter(MetaMap& metadata);
+    explicit StringMetaInserter(MetaMap& metadata);
+
+    /// Returns @c true if key exists
+    bool hasKey(const Name& key) const;
+    /// Returns @c true if index exists
+    bool hasIndex(Index index) const;
 
-    /// Insert the string into the metadata
-    void insert(const Name& name);
+    /// Insert the string into the metadata, hint is used if non-zero
+    /// Returns the chosen index which will match hint if not in use
+    Index insert(const Name& name, Index hint = Index(0));
 
     /// Reset the cache from the metadata
     void resetCache();
 
 private:
+    using IndexPairArray = std::deque<std::pair<Index, Index>>;
```

Comment:
why deque? I'd imagine in the general case a vector is probably more efficient? Unless your indices vary widely for a large number of points

Reply:
Oh nvm I forgot my own algorithm... I agree with deque